persist: yield to timely while awaiting `Future`s

petrosagg opened the pull request:
### Motivation

The only way to use async futures inside a timely worker currently is to block the whole timely thread while that `Future` is pending. As more and more of our persist story revolves around rust async futures (e.g while talking to S3) the blocking problem will become worse.

This PR introduces an `OperatorBuilder::build_async()` method that allows supplying the operator logic as an async future instead of a closure which allows for arbitrary `await`ing on futures.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR adds a release note for any
  [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

ruchirK commented on src/persist/src/operators/mod.rs:
```diff
@@ -24,3 +39,121 @@ pub(crate) fn split_ok_err<K, V>(
+    }
+}
+
+trait OpeatorBuilderExt<G: Scope> {
```
nit: Operator

petrosagg replied:
oops, I don't know how I manage to mispell it three times :P fixed

ruchirK commented on src/persist/src/operators/mod.rs:
```diff
@@ -24,3 +39,121 @@ pub(crate) fn split_ok_err<K, V>(
+                Scheduler(Rc::clone(&scheduled)),
+            ));
+            move |frontiers| {
+                // If the future is currently borrowing the shared frontier we'll just go ahead and
```
nit: it seems like you poll the future regardless of whether its currently borrowing the shared frontier or not in line 142, but the way the comment is worded it sort of seems like we only poll the future iff its currently borrowing the shared frontier?

petrosagg replied:
I'll rephrase the comment but the code is correct. This RefCell borrow can fail if the future also borrowed the frontier internally and kept the borrow alive across an await point. So it is crucial that we poll the future anyway so that it progresses through its state machine and eventually, in some future invocation of the operator, the lock will be released (because the future would have reached the `scheduler.yield_now().await` point

petrosagg replied:
@ruchirK I rephrased the comment, let me know if's clear now

ruchirK replied:
Yeah the new comment is great! Super minor nit: I think you mean "gets rescheduled" instead of "gets rescheduler", but don't bother fixing that unless you have to go through ci again

ruchirK replied:
Trust me I had no doubts the code was correct :)

petrosagg replied:
Fixed the typo :) I don't have sympathy for the CI working hard, it can rebuild the whole thing to change a letter :P

ruchirK reviewed (commented):
I don't fully understand this yet (due to my own ignorance not a fault in the code) but this is cool!

danhhz commented on src/persist/src/operators/stream.rs:
```diff
@@ -269,16 +270,17 @@ where
 
                 if !active_seal_operator {
-                    return;
+                    continue;
```
initial question while I'm digesting the actual implementation bits: if this is expected to always be used in a `while scheduler.yield_now().await` loop, can we push that loop into build_async (or offer a variant of build_async that does it)? it seems somewhat idiomatic in rust for macros to hide early returns (e.g. `try!/?`, `bail!`, `ready!`) and if we can, we may as well lean into that instead of having to be careful about only using `continue`s

petrosagg replied:
I've been grappling with this too, I don't like the `continue` here at all. There is one ugly solution of doing this:

```rust
while scheduler.yield_now().await {
    async {
        if early_return_condition {
            return;
        }
    }.await;
}
```

Which allows us to use `return`, `bail!` etc but increases even more the boilerplate. Hiding these details inside `build_async` I think is going to make the API horrible for the following reasoning:

1. If `build_async` itself builds the future containing the while loop, then it must be given a logic function that can produce the inside future on demand like so: ` async { while ... { logic().await } }`
2. Since that logic function must be callable in a loop it cannot be an `FnOnce`, it must be an `FnMut` or an `Fn`
3. The useful futures that we want to return from this logic function will almost certainly be capturing variables from that closure's environment. This is not allowed and the compiler will complain that closure state is escaping. Then the only solution for the API user could be to use some kind of `Rc<RefCell<..>>` for every single captured closure state, and clone the, every single time they produce a future.

I'll keep ruminating on this in case a better idea comes along

petrosagg replied:
Here is the error I'm talking about in isolation: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d3774452f48e02cc95dc55f1771bdb5c

danhhz replied:
Yeah that's compelling. If we stick with the while loop, I actually prefer the additional boilerplate. I'll take "clunky but easy to reason about" over "more elegant but more subtle" any day

This is a shot in the dark, but would it work to use a 2-level closure to solve the Fn ownership issues?

danhhz replied:
> This is a shot in the dark, but would it work to use a 2-level closure to solve the Fn ownership issues?

Ah I see, maybe not. This works for timely operators because the logic closure gets called in a loop, but those communicate by emitting things not by returning them, so we still have a (3) issue

(Whoops, wrote this in the wrong place the first time.)

petrosagg replied:
> I actually prefer the additional boilerplate. I'll take "clunky but easy to reason about" over "more elegant but more subtle" any day

Definitely agree! I hadn't thought of the extra boilerplate solution until your comment so thanks for the nudge :) I'll change the PR to that style

Btw, the real solution is to not rely on closures to provide the logic but rather define an `AsyncOperatorLogic` trait that captures this lifetime relationship and provide an `L: AsyncOperatorLogic` struct to build_async. Making timely async-friendly is a goal  I'm making slow progress on so we might see that in the future

danhhz commented:
> This is a shot in the dark, but would it work to use a 2-level closure to solve the Fn ownership issues?

Ah I see, maybe not. This works for timely operators because the logic closure gets called in a loop, but those communicate by emitting things not by returning them, so we still have a (3) issue

danhhz commented on src/persist/src/operators/mod.rs:
````diff
@@ -24,3 +39,125 @@ pub(crate) fn split_ok_err<K, V>(
+/// }
+/// ```
+#[derive(Debug)]
+pub struct Scheduler(Rc<Cell<bool>>);
````
nit: it might be a touch easier to read all this if this field was named (scheduled? yielded?)

danhhz commented on src/persist/src/operators/mod.rs:
```diff
@@ -24,3 +39,125 @@ pub(crate) fn split_ok_err<K, V>(
+                }
+
+                let was_marked_scheduled = scheduled.replace(true);
+                let _ = Pin::new(&mut logic_fut).poll(&mut Context::from_waker(&waker));
```
this only returns a Ready if the logic violates the contract ("The logic constructror is expected to return a future that will never return") right? it might be worth logging or debug_assert'ing or something to make it more obvious what is going on if someone accidentally makes a change that violates that contract

danhhz reviewed (approved):
I'm a little over my head in terms of necessary future/async/timely background to truly convince myself of the correctness of this (my last company called this sciencedog in apparent reference to https://knowyourmeme.com/memes/i-have-no-idea-what-im-doing). I'll take the time to brush up on that at some point, who knows when that will be, so don't block merging the PR on it

petrosagg commented:
@danhhz this should be ready to merged now. I just pushed these changes:

1. The module has been moved to `async_ext.rs` (plain async is a keyword)
2. Added a simple test that makes sure the operator works
3. Introduced an `async_op!` macro that hides the whole `loop { yield_now().await` thingy and allows using `return` in the operator logic normally
4. Enforced with the type system that the `Future` never returns by creating a `Never` type that is not inhabited. It works like the unstable `never` type of std
5. Added fields and methods to `Scheduler` to make them easier to read

danhhz commented:
lovely, thank you!
